main.go

```diff
@@ -190,17 +211,20 @@ func main() {
 		os.Exit(1)
 	}
 
-	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up health check")
-		os.Exit(1)
-	}
-	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up ready check")
-		os.Exit(1)
-	}
+	leaderElectionContext, cancelLeaderElection := context.WithCancel(context.Background())
```

Comment:
That way `leaderElectionContext` would stop as soon as the stop context is Done. Which is what I'm trying to achieve with this PR.

I tried using nested contexts here
https://go.dev/play/p/6dFfBQyXlW1

Reply:
by using the stop context as parent of this one, the leaderElectionMgr.Start function will be cancelled if a singint or sigterm is received.

Reply:
Yes, that's what I don't want to happen. Here's the ending sequence

- sigint arrive
- stopCh is Done
- mgr and mgrGlobals finish their work and return
- as mgr is not on a go routing, when it returns it complete the function, triggering all the `defer`s
- `utils.Shutdown()` defer goes first. do the cleanup (the leader election is still running here, so we have the lock lease)
- `cancelLeaderElection()` defer goes second, stopping the manager which in turn release the lock

Am I missing something here?

Reply:
Right, it would be the same as using the `ReleaseOnCancel` option.

My concern here is that after calling `cancelLeaderElection()` (with the defer), the program exits, so nothing ensures that `leaderElectionMgr` really finishes. There is a race condition.

The internal implementation of the `ReleaseOnCancel` option uses a channel to ensure that the routine has stopped before continuing with the shutdown. Perhaps you can do the same here.

Reply:
> My concern here is that after calling cancelLeaderElection() (with the defer), the program exits, so nothing ensures that leaderElectionMgr really finishes. There is a race condition.

Added a wait group to ensure leaderElectionMgr is correctly stopped

Reply:
it seems to me that this never exits. The leader election manager is stopped with a defer but before that, you are waiting for this to stop, so, it won't happen?

Reply:
fixed